src/extendedstorage.cpp

```diff
@@ -729,53 +767,56 @@ void ExtendedStorage::Private::clearAlarms(const QString &notebookUid)
 
 void ExtendedStorage::Private::setAlarmsForNotebook(const Incidence::List &incidences, const QString &notebookUid)
 {
-    const QDateTime now = QDateTime::currentDateTime();
-    // list of all timed events
-    Timed::Event::List events;
-    foreach (const Incidence::Ptr incidence, incidences) {
-        setAlarms(incidence, notebookUid, events, now);
+    MemoryCalendar::Ptr calendar(new MemoryCalendar(QTimeZone::utc()));
```

Comment:
What's the reason for adding a MemoryCalendar?

Reply:
In this `setAlarmsForNotebook()`, used when switching notebook visibility, I've the same issue. I can't directly call `addAlarms()` for every incidence with `now` as a starting point in time to get alarms because recurring incidences may have exceptions in the list. To tackle this, I've two possibilities:
- scan the list argument to fetch recurring incidences and their exceptions and mimic what is done in `setAlarms()` before calling `addAlarms()`,
- or reuse `setAlarms()` itself, which requires a `Calendar` to be able to call `Calendar::instances()`.

I chose the second options because it requires less lines, duplicate less code, and `MemoryCalendar` is actually a way to achieve the first options with a low overhead. Do you think it's unwise ? I've added lines in the `tst_recurringAlarms()` test specifically for this. And it doesn't break `tst_alarms()` neither.

While thinking about this, I'm even wondering if I'm not going to change `ExtendedStorage::allIndiences()` and friends to return a `MemoryCalendar` directly, in the API breaking PRs. Indeed, at the moment, these routines are very fragile because they are returning a list of incidences by scanning the DB, but without associating incidence and notebook (most of the time they are called with a notebook argument anyway which mitigate the issue of course). But that's another story…

Reply:
> what would go wrong if instead of the separately created MemoryCalendar the main calendar was passed to setAlarms(). Is ::instances(incidence) not working the same with that?

Actually no, it won't work because `ExtendedStorage::allIncidences()` is returning incidences not belonging to any calendars. It is returning a list created from scratch (well created from incidences created on the fly by reading the DB).

> Since the name is "for notebook" the method could internally fetch the list it needs to update?

That would be nice, but I would like to keep all the alarm code away from `ExtendedStorage`. I mean, in a way that all their arguments are not containing any pointer to an `ExtendedStorage`, so one day this code could be moved away from `ExtendedStorage` itself. Decorrelated somehow. That's why I don't like the `allIncidences()` current API because when you want to use its result, you always need to pass the incidence list and the notebook id it is associated to (without speaking of the mess it could be if you have called it without notebook id argument). That's why `setAlarmsForNotebook()` have the two arguments. It would be nicer to have `setAlarmsForIncidences(const MemoryCalendar::Ptr &calendar)`. If you prefer, I can rename the routine and create the memory calendar in `updateNotebook()` instead.

Reply:
> Actually no, it won't work because ExtendedStorage::allIncidences() is returning incidences not belonging to any calendars. It is returning a list created from scratch (well created from incidences created on the fly by reading the DB).

That's getting somewhere. Perhaps the belonging to calendars is not problem yet per se but doing things like Incidence::List::contains().

> That would be nice, but I would like to keep all the alarm code away from ExtendedStorage

Suppose that means now that the stuff implemented in ExtendedStorage::Private don't have any reference to the parent. And indeed the private class doesn't even have a pointer to it, making it impossible to simply call the allIncidences() there.

As such I'm fine with this.

Reply:
I've removed the setAlarmsForNotebook() routine and move the code o the updateNotebook() one. Like that the creation of the calendar is close to the allIncidences() call.